lib/automation.go

```diff
@@ -251,22 +252,6 @@ func deploy(scope scope, p *DeployParams) {
 	log.Debugw("deploy started", "ref", vi.SimpleRef().String(), "payload", deployPayload)
 	scope.sendEvent(event.ETAutomationDeployStart, ref, deployPayload)
 
-	rollback := true
-	defer func() {
```

Comment:
Moving this defer function later means it won't run if an error & return happens before it gets reached. That doesn't currently matter because if `Save` fails there's nothing to `Remove`, but it makes it easier to introduce bugs later on if this code path changes. Generally speaking, defer'd rollbacks should be setup early, such that they handle all failure cases that might occur.

Rather, to achieve the effect you're going for, I think it would be better to add `changesSaved := false` on the line above (which reads as "no changes have been saved yet"), keep your `if rollback && changesSaved` condition, and then after `Save` happens, make that assignment into: `changesSaved = !errors.Is(err, dsfs.ErrNoChanges)`

Reply:
good point! thought I was being *clever* lol